organizations/data.py

```diff
@@ -327,39 +353,64 @@ def bulk_create_organization_courses(organization_course_pairs):
         in organization_course_pairs
     }
 
-    # Grab all existing (lowercased org short name, course key string) pairs from db,
-    # filtering for the ones requested for creation in `organization_course_pairs`, and
-    # indexing by db `id` of org_course object (we need this later for the bulk update).
-    orgslug_courseid_pairs_to_activate_by_id = {
-        org_course_id: (short_name.lower(), course_id)
-        for org_course_id, short_name, course_id
+    # Grab all existing (lowercased org short name, course key string, is_active) triples
+    # from db, filtering for the ones requested for creation in `organization_course_pairs`,
+    # and indexing by db `id` of org_course object (we need this later for the bulk update).
+    existing_orgslug_courseid_triples_by_id = {
+        org_course_id: (short_name.lower(), course_id, is_active)
+        for org_course_id, short_name, course_id, is_active
         in internal.OrganizationCourse.objects.values_list(
-            "id", "organization__short_name", "course_id"
+            "id", "organization__short_name", "course_id", "active"
         )
         if (short_name.lower(), course_id) in orgslug_courseid_pairs
     }
-
-    # Working backwards from the set of pairs we need to *activate*,
-    # find the set of pairs we need to *create*.
-    orgslug_courseid_pairs_to_activate = set(
-        orgslug_courseid_pairs_to_activate_by_id.values()
+    existing_orgslug_courseid_triples = set(
+        existing_orgslug_courseid_triples_by_id.values()
     )
+
+    # Now that we have the set of orgslug/courseid linkages that *already exist*,
+    # determine which ones need to be *reactivated*
+    # and which linkages from our original set we still need to *create*.
     orgslug_courseid_pairs_to_create = {
         (orgslug, courseid)
         for orgslug, courseid in orgslug_courseid_pairs
-        if (orgslug, courseid) not in orgslug_courseid_pairs_to_activate
+        if not (
+            # If an organization-course linkage already exists as either
+            # active OR inactive, we do not need to create it.
+            (orgslug, courseid, True) in existing_orgslug_courseid_triples or (
+                (orgslug, courseid, False) in existing_orgslug_courseid_triples
+            )
+        )
```

Comment:
Nit: This is pretty tricky to follow. The comments help, but it might be clearer to make a small helper function (e.g. `link_exists`).

Reply:
Good point. I didn't do exactly that, but I tried to bring the readability up a lot.